certs/certs.go

```diff
@@ -0,0 +1,293 @@
+	totalPower := new(gpbft.StoragePower)
+	for i := range powerTable {
+		totalPower = totalPower.Add(totalPower, powerTable[i].Power)
+	}
+
+	signers := make([]gpbft.PubKey, 0, len(powerTable))
+	signerPower := new(gpbft.StoragePower)
+	if err := cert.Signers.ForEach(func(i uint64) error {
+		if i >= uint64(len(powerTable)) {
+			return xerrors.Errorf(
+				"finality certificate for instance %d specifies a signer %d but we only have %d entries in the power table",
+				cert.GPBFTInstance, i, len(powerTable))
+		}
+		signer := &powerTable[i]
+		signerPower = signerPower.Add(signerPower, signer.Power)
+		signers = append(signers, signer.PubKey)
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	if !gpbft.IsStrongQuorum(signerPower, totalPower) {
+		return xerrors.Errorf("finality certificate for instance %d has insufficient power: %s < 2/3 %s", cert.GPBFTInstance, signerPower, totalPower)
+	}
+
+	payload := &gpbft.Payload{
+		Instance:         cert.GPBFTInstance,
+		Round:            0,
+		SupplementalData: cert.SupplementalData,
+		Step:             gpbft.DECIDE_PHASE,
+		Value:            cert.ECChain,
+	}
+
+	// We use SigningMarshaler when implemented (for testing), but only require a `Verifier` in
+	// the function signature to make it easier to use this as a free function.
+	var signedBytes []byte
+	if sig, ok := verifier.(gpbft.SigningMarshaler); ok {
+		signedBytes = sig.MarshalPayloadForSigning(nn, payload)
+	} else {
+		signedBytes = payload.MarshalForSigning(nn)
+	}
+
+	if err := verifier.VerifyAggregate(signedBytes, cert.Signature, signers); err != nil {
+		return xerrors.Errorf("invalid signature on finality certificate for instance %d: %w", cert.GPBFTInstance, err)
+	}
+	return nil
+}
+
+// MakePowerTableDiff create a power table diff between the two given power tables. It makes no
+// assumptions about order, but does assume that the power table entries are unique. The returned
+// diff is sorted by participant ID ascending.
+func MakePowerTableDiff(oldPowerTable, newPowerTable gpbft.PowerEntries) []PowerTableDelta {
+	oldPowerMap := make(map[gpbft.ActorID]*gpbft.PowerEntry, len(oldPowerTable))
+	for i := range oldPowerTable {
+		e := &oldPowerTable[i]
+		oldPowerMap[e.ID] = e
+	}
+
+	var diff []PowerTableDelta
+	for i := range newPowerTable {
+		newEntry := &newPowerTable[i]
+		delta := PowerTableDelta{ParticipantID: newEntry.ID}
+		if oldEntry, ok := oldPowerMap[newEntry.ID]; ok {
+			delete(oldPowerMap, newEntry.ID)
+			delta.PowerDelta = new(gpbft.StoragePower).Sub(newEntry.Power, oldEntry.Power)
+			if !bytes.Equal(newEntry.PubKey, oldEntry.PubKey) {
+				delta.SigningKey = newEntry.PubKey
+			}
+			if delta.IsZero() {
+				continue
+			}
+		} else {
+			delta.PowerDelta = newEntry.Power
+			delta.SigningKey = newEntry.PubKey
+		}
+		diff = append(diff, delta)
+	}
+	for _, e := range oldPowerMap {
+		diff = append(diff, PowerTableDelta{
+			ParticipantID: e.ID,
+			PowerDelta:    new(gpbft.StoragePower).Neg(e.Power),
```

Comment:
There's a subtle point here that a power delta that takes the result to zero is to be interpreted as a deletion. Deletion of the signing key isn't explicit here. I think it's worth documenting this interpretation, perhaps with the PowerTableDelta type. Or through the code with a function like `newPowerDeltaDelete(id, e.power)`

Reply:
Fixed in #292, #291.